### PR TITLE
refactor: project-wide audit of helper/util usage in src/ + OCPP 2.0.1 §F01.FR.13 spec fix

### PR DESCRIPTION
**Pull request by jerome-benoit**
## Highlights

This PR is **two changes in one**:

1. **OCPP 2.0.1 §F01.FR.13 spec compliance fix** (most consequential). Removes fabricated `transactionId: generateUUID()` from 11 `RequestStartTransaction` rejection sites and adds the cable-plugin-first echo branch (`TxStarted` reasonCode + existing `connectorStatus.transactionId`) per §F01.FR.13. The simulator becomes the only OSS OCPP 2.0.1 impl found that honors this SHALL (lorenzodonini/ocpp-go, mobilityhouse/ocpp, citrineos-core, EVerest/libocpp all defer it).
2. **Helper/util adoption audit across `src/`** (original PR mandate). 75+ sites migrated to `getErrorMessage`, `clone`, `convertToBoolean`, `isEmpty`/`isNotEmptyString`, `convertToInt`/`has`, `isNotEmptyArray`, `handleIncomingRequestError`, `JSONStringify`, `assertIsJsonObject`, `promiseWithTimeout`, `ensureError`. Plus 9 error subclasses gained nominal-type discriminants (`public override readonly name = '...' as const`), 1 file-private helper (`buildRejectedResponse`), 1 sentinel error class (`TimeoutError`).

**Side-effect bug fixes uncovered during the audit:**
- `Configuration.ts`: `convertToInt(env.PORT)` crashed on unset PORT (empty string).
- `getErrorMessage`: previously emitted meaningless output for non-`Error` thrown values.
- `JSONStringify`: silently dropped `Map` entries.
- `TemplateSchema.ts`: non-integer / negative EVSE keys silently accepted (no Zod issue).
- 4 OCPP spec citation corrections (§E07→§F01.FR.13, §7.2→§7.1, `InvalidValue`→`InvalidProfile` for F01.FR.11, citation qualifier coherence).

**Merge note**: 22 commits include 6 "apply vN cross-validation review fixes" bundles. **Squash-merge recommended**; suggested squashed-commit title: `refactor: project-wide audit of helper/util usage in src/ + OCPP 2.0.1 §F01.FR.13 spec fix`.

---

## PR Checklist

- [x] Please add a description of your changes.
- [x] Please ensure title follows [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] We need your changes to come with unit tests in order to keep this project in quality and easy to maintain.
- [x] If your changes contain any new feature please be sure to document it. Not applicable: refactor only, no public API or behaviour change.
- [ ] Please add a link to the open issue or task that this pull request will resolve. Not applicable: internal audit.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

---

## Description

Project-wide audit of helper/util usage in `src/` (excluding `tests/`, which was covered by #1930). 16 commits, 21 files, +447/-353 net, 75 sites refactored across two audit passes + four review rounds (v1/v2/v3/v4) + one external bot review.

### Pass 1 — predicates / conversions / OCPP-error / log / config helpers

**Correctness fixes**
- `(error as Error).message` → `getErrorMessage(error)` (3 sites): unsafe cast silently emits `"undefined"` for non-`Error` thrown values, masking crypto/signing failures.
- bare `structuredClone(x)` → `clone(x)` (7 sites): DRY consolidation onto the existing wrapper.

**Correctness-latent fixes**
- `?.value === 'true'` → `convertToBoolean(...)` (7 sites): the inline pattern only matches the literal lowercase `'true'`; `convertToBoolean` handles `'True'`, `'1'`, native `true`, whitespace, etc. Plus extracted local helpers — `isOCPP20FlagEnabled` in `OCPPServiceUtils.ts` (3 sites collapsed) and `isSigningStartedReadingsEnabled` / `isSigningUpdatedReadingsEnabled` siblings of `isSigningEnabled` in `OCPP16ServiceUtils.ts`.
- `.length === 0` / `.size === 0` on typed collections → `isEmpty(...)` (7 sites).
- `.length > 0` / `== null || .length === 0` on strings → `isNotEmptyString(...)` (2 sites).
- `Number.parseInt(_, 10)` / `Number(_)` on unknown input → `convertToInt(...)` (2 sites; a third was reverted to `Number(value)` after a `hyperspace-insights[bot]` review uncovered 4 divergence cases including uncaught exceptions and silent truncation — see commit `893968ad1`).
- `Object.hasOwn(obj, prop)` → `has(prop, obj)` (2 sites).

**Cosmetic**
- `.length > 0` on typed array → `isNotEmptyArray(...)` (4 sites).
- inline `${component}.${variable}` log string → `buildConfigKey(component, variable)` (1 site).

### Pass 2 — less-obvious helpers (timers, JSON, OCPP error routing, type guards)

- `handleIncomingRequestError<T>` (10 single-catch handlers in `OCPP20IncomingRequestService.ts`): brings OCPP 2.0 to parity with OCPP 1.6 for the simple `catch → return typed-rejection` pattern.
- `JSONStringify(_, 2 [, MapStringifyFormat.object])` (14 sites): replaces bare `JSON.stringify(_, undefined/null, 2)`. The bare form silently drops `Map` entries; the helper serializes them safely. `MapStringifyFormat.object` is used wherever the value may carry Maps. The `JSONStringify` generic constraint was loosened to `object: unknown` to align with `JSON.stringify`'s standard signature.
- `assertIsJsonObject(value, error?)` (2 sites): replaces inline guard, narrows the value type to `JsonObject`, preserves the existing `BaseError` message + stack.
- `promiseWithTimeout(promise, ms, error)` (1 site): replaces a manual `new Promise(...)` + `setTimeout` reject + `clearTimeout` construct.

### V3 cross-validated fixes (commit `564c74909`)

- **CV1** — `isOCPP20FlagEnabled` parameter tightened to OCPP 2.0 variable-name union.
- **CV2** — `convertToInt(evseKey)` in `TemplateSchema.ts` reverted to `convertToIntOrNaN(evseKey)` (preserves NaN-on-invalid contract the constraint check relies on).
- **CV3** — `error === timeoutError` identity check replaced with `instanceof TimeoutError` against a `BaseError` subclass discriminator.
- **CV4** — 4 remaining `handleIncomingRequestError` sites in OCPP 1.6 aligned to the local-const pattern OCPP 2.0 adopted.
- **S1 (v3)** — reverted 4 nested + outer catches in OCPP 2.0 `handleRequestStartTransaction` to inline `catch → logger.error → return typed-rejection` for observability.
- **S4-S7** — README `SIMULATOR_COLD_START` doc, Bootstrap.ts `MapStringifyFormat.object` on worker log, TemplateSchema TDZ comment declarative form, OCPP16 isSigning*ReadingsEnabled JSDoc terminology.

### V4 cross-validated fixes (commit `0e70bc4b8`)

**Correctness:**
- **BUG** — `Configuration.ts:293`: `convertToInt(env.PORT ?? '')` would throw on empty string when `PORT` is unset. Same class of bug as CV2. Now guarded with `isNotEmptyString(env.PORT)`.

**Design:**
- **HC1 + Q2(d)** — `TimeoutError` relocated from `Bootstrap.ts` to a dedicated `src/exception/TimeoutError.ts` file, exported from `src/exception/index.js`. Matches the BaseError/OCPPError pattern. No longer interleaved with imports.
- **S1 (v4)** — extracted `rejectedInternalError(additionalInfo)` helper in OCPP20IRS. Collapses 4 near-identical 8-line typed-response literals at the `handleRequestStartTransaction` catches to 1-line calls.
- **S2** — wrapped `logger.error(..., error)` with `ensureError(error)` at the 4 G-sites for cross-site consistency with the 10 sibling sites.
- **Q3(a)** — added `Number.isNaN(evseId)` guard + `ctx.addIssue` in `TemplateSchema.ts` `.superRefine` loop over Evses. Closes the silent validation gap for non-numeric Evse keys.

**Polish:**
- **S3** — `Bootstrap.ts:729`: `logger.warn(..., error.message)` instead of closed-over `timeoutError.message` after `instanceof TimeoutError` narrowing.
- **S4** — `behaviour` → `behavior` (cspell + repo convention).
- **S5** — catch comment compressed from 3 lines to 1; removes the `unlogged` cspell flag and the partial redundancy with `TimeoutError` JSDoc.
- **HC2** — README `SIMULATOR_COLD_START` wording: `"case-insensitive true or 1, optionally surrounded by whitespace"` (exhaustive rule, not non-exhaustive example list).

### V5 cross-validated fixes (commit `23beb7ea1`)

**Spec correctness:**
- **L-A + HC-D** — OCPP 2.0.1 §F01.FR.13 spec deviation in `RequestStartTransaction` rejection responses. Per §F01.FR.13, the Charging Station SHALL return `transactionId` in `RequestStartTransactionResponse` only when a transaction was already created on the station but not yet authorized (cable-plugin-first); on pure rejections it MUST NOT be fabricated — doing so misleads CSMS implementations that map `remoteStartId → transactionId`. Renamed the v4 helper `rejectedInternalError(additionalInfo)` to `buildRejectedResponse(additionalInfo, reasonCode)` and migrated all 11 rejection sites in `handleRequestStartTransaction` (4 prior helper calls + 7 inline literals: TxInProgress, MasterPassGroupId InvalidIdToken, IdToken not authorized, GroupIdToken not authorized, InvalidProfile purpose, InvalidValue, InvalidProfile invalid) through it. Dropped `transactionId: generateUUID()` from all 11. Inline JSDoc anchors the rationale to §F01.FR.13.

**Polish:**
- **HC-A** — `TemplateSchema.ts:271`: replaced v4's `Number.isNaN(evseId)` arithmetic-NaN guard with `!Number.isInteger(evseId) || evseId < 0` for explicit intent.
- **HC-B** — `TimeoutError.ts`: JSDoc compressed to single line.
- **HC-C** — `TimeoutError.ts`: SPDX copyright header added to match peer `OCPPError.ts`.
- **V5-6** — `Bootstrap.ts`: catch-block ordering comment relocated below `instanceof` check and reworded for clarity.

**Test:**
- `OCPP20IncomingRequestService-RequestStartTransaction.test.ts:434`: §F01.FR.13 regression assertion updated to expect `response.transactionId === undefined` on TxInProgress rejection (was incorrectly asserting `notStrictEqual` against the now-fixed bug). Inline §F01.FR.13 anchor prevents future reverts.

**Deferred from v5 single-agent findings:**
- **EX-A** — Two other `promiseWithTimeout` sites use plain-string error messages instead of `TimeoutError` (pre-existing pattern, out of scope).
- **V5-5** — `env.PORT` empty-string fallback now logs warn via CF logger (new behavior introduced in v4 BUG fix; documented but no additional change).
- **V5-8** — README "case-insensitive" wording (pedantic, v4 HC2 wording retained).
- **OCPP20VariableManager NaN comparisons** (Librarian v5 finding): pre-existing, out of scope.

### V5.1 spec citation correction (commit `de9a78500`)

The L-A/HC-D fix in `23beb7ea1` originally cited `§E07`, inherited from a pre-existing in-code comment. Verification against `OCPP-2-0-1-edition3-part2-specification.md` (local spec collection) showed `§E07` is "Transaction locally stopped by IdToken" — unrelated. The actual governing clause is **§F01.FR.13** (chapter F. RemoteControl, use case F01 - Remote Start Transaction - Cable Plugin First): _"When a transaction is created on the Charging Station, but has not been authorized. AND RequestStartTransactionRequest is received → The Charging Station SHALL return the transactionId in the RequestStartTransactionResponse."_

Semantics of the fix are unchanged — only the spec citation in the helper comment, test comment, and this PR description is corrected.

### V6 cross-validation review (4 agents, commit `9374dd39e`)

Re-applied the full 4-agent method (Oracle ×2, Librarian, Explore) with two session-level criteria added: (1) every OCPP spec citation MUST be qmd-verified against the local `ocpp-specs` collection, (2) the v5 `transactionId`-drop fix audited for the cable-plugin-first edge case.

**Cross-agent HIGH convergence:**
- **V6-S2 + V6-Q1** (both Oracles) — RequestStartTransaction `TxInProgress` rejection regressed §F01.FR.13. When `connectorStatus.transactionPending === true` and `connectorStatus.transactionId` was a string (cable-plugin-first scenario), the v5 helper unconditionally dropped the **real** transactionId. v5 fixed the fabricated-UUID bug but went too far. Fix: split the check into two branches — the `transactionPending`-with-existing-id branch now echoes `connectorStatus.transactionId` via the extended helper, with `ReasonCodeEnumType.TxStarted` (appendix-correct: TxStarted = cable-plugin-first, not yet authorized; TxInProgress = authorized and running). The TxInProgress fallthrough remains for `transactionStarted || locked` cases. Helper signature extended: `buildRejectedResponse(reasonCode, additionalInfo, transactionId?)`.

**qmd-verified citations:**
- V6-S1 (M) — `TemplateSchema.ts` cited §7.2 (Connector numbering) for EVSE keys. qmd against `OCPP-2.0.1_edition3_part1_architecture_topology.md:269` confirms §7.1 is EVSE numbering, §7.2 is Connector numbering. Two occurrences corrected.
- V6-E3 (Explore HIGH flag) — **false positive**. `B09.FR.31 (errata 2025-09 §2.12)` is genuinely listed in `OCPP-2.0.1_edition3_errata_2025-09.md:30` p.64 §2.12. Citation valid.
- V6-E4 (Explore M flag) — **false positive**. §2.10 is the right ChargingProfileType datatype section per `part2_specification.md:6155`; the transactionId row explicitly carries the "SHALL only be included when ChargingProfilePurpose is set to TxProfile in a SetChargingProfileRequest" constraint.

**Other in-scope fixes:**
- V6-S4 (L, Oracle1) — F01.FR.11 violation (ChargingProfile.transactionId set in RequestStart) now uses `InvalidProfile` reasonCode (was `InvalidValue`) for family coherence with the other 2 ChargingProfile rejection sites per F01.FR.26 hint.
- V6-X1 (L, Librarian) — Helper comment overstated §F01.FR.13 by treating the silent-on-rejection case as a SHALL NOT. Reworded to distinguish the positive obligation (SHALL echo when precondition holds) from the practical argument (fabrication misleads CSMS).
- V6-X3 + V6-Q11 (L, Librarian + Oracle2) — Swapped `buildRejectedResponse` argument order to `(reasonCode, additionalInfo, transactionId?)` — required-first/optional-last, matches schema field importance. All 11 existing callsites updated; V6-S2 branch is the 12th call.
- V6-X14 (M, Librarian) — `TimeoutError extends BaseError {}` was an empty subclass; TypeScript structural typing made BaseError/TimeoutError indistinguishable in the type system, leaving the `instanceof` check in `Bootstrap.ts` dependent on the runtime prototype chain only. Added `public override readonly name = 'TimeoutError' as const` for nominal distinction (matches `OCPPError` pattern).
- V6-Q2 (M, Oracle2) — Test coverage gap: only 1 of 11 v5 rejection paths asserted `transactionId === undefined`. Added the assertion to the 6 remaining unprotected Rejected tests. The V6-S2 fix test now asserts `transactionId === pendingTransactionId` AND `reasonCode === TxStarted` (regression-locks the §F01.FR.13 echo path).

**Deferred to follow-up PRs (documented):**
- V6-Q3 — Mixed catch pattern in `handleRequestStartTransaction` (4 sites use local `logger.error + buildRejectedResponse`, others go through `handleIncomingRequestError`). Intentional per v3 S1 decision for per-phase observability.
- V6-E11 + V6-E12 + V6-E15 — `ensureError()` missing at 18 `.catch((error: unknown) => { logger.error(..., error) })` sites in OCPP 2.0 and 2 OCPP 1.6 catches (`handleRequestGetLocalListVersion`, `handleRequestSendLocalList`). Pre-existing systematic gap, out of audit scope.
- V6-E19 — `handleRequestStopTransaction` has 3 inline rejection literals that could use a typed helper (different return type from `buildRejectedResponse`). Requires generic refactor.
- V6-X4/X7/X12/Q4/Q5 — JSDoc length note, generic constraint loosening trade-off, stale §E07 in v5 commit body (git history immutable), redundant `errorResponse` type annotations, Bootstrap.ts comment trim.
- V6-E1/E2/E5/E6 — OCPP 1.6 bare `§N.NN` citations lacking version qualifier (cosmetic).

### V7 cross-validation review (4 agents, commit `4d870dd85`)

Re-applied the method with v6 lessons elevated to non-negotiable criteria: regression-chain audit (V6-S2 was introduced by v5), type-cast scrutiny (`as UUIDv4` justified by runtime evidence), test-gap blast-radius (audit ALL sibling test files), helper-signature swap audit, mandatory qmd-verification of every spec citation before treating as a blocker (lesson V6-E3/E4 false positives).

**False positive rejected (qmd + caller-trace verification):**
- **V7-E2/E17** (Explore HIGH unique-agent) — claimed `OCPP20ServiceUtils.startTransactionOnConnector` is missing `transactionPending = true` assignment. Rejected after tracing callers: the function is invoked only by the Automatic Transaction Generator with `triggerReason: Authorized` — it's the authorized-start path, not cable-plugin-first. Setting `transactionPending = true` there would create spurious §F01.FR.13 echoes on already-authorized connectors. Same-class false positive as V6-E3 (B09.FR.31).

**Convergent findings applied:**
- **V7-Q1 + V7-X5** (Oracle2 + Librarian) — Conditional spread idiom: swapped `?: {}` → `&& { x }` to match the dominant codebase pattern (verified across `ChargingStation.ts`, `OCPP16TestUtils.ts`, `AbstractUIService.ts`).
- **V7-Q4 + V7-S1** (Oracle2 + Oracle1) — Helper docstring trimmed from 6 lines to 3 + reworded to avoid implying §F01.FR.13 mandates `status = Rejected` (the SHALL is silent on status; the v6 phrasing read as if Rejected were mandated).
- **V7-Q5** (Oracle2) — Test §F01.FR.13 echo comment was P4 (verbatim duplicate of helper docstring); replaced with 1-line anchor `// §F01.FR.13: echo pending transactionId`. Assertions already self-document.
- **V7-Q7 + V7-S2 + V7-X6** (all 3 agents) — Added inline safety comment at the `connectorStatus.transactionId as UUIDv4` cast site pointing to the source-of-truth write site (line ~2740 `generateUUID()` assignment).
- **V7-Q2** (Oracle2) — Added `statusInfo.reasonCode` assertions to 6 v6-Q2 tests so a future bug swapping reasonCodes cannot pass: `InvalidIdToken`×2, `InvalidProfile`×2, `InternalError`×1, `NotFound`×1.
- **V7-Q3** (Oracle2) — Added test exercising the residual `TxInProgress` branch via `connectorStatus.locked = true` — verifies `reasonCode = TxInProgress` and `transactionId = undefined` (the V6-S2 split's fallthrough branch was previously untested).
- **V7-S6 + V7-E4** (Oracle1 + Explore) — V6-Q2 spillover gap closed: added `transactionId === undefined` + `reasonCode = InvalidIdToken` assertions to `MasterPass.test.ts:112`.
- **V7-E5** (Explore) — Added `reasonCode = InvalidValue` ×3 assertions to the 3 `RequestStopTransaction` invalid-UUID-format rejection tests. (All 3 hit the format-check at line 2783 before the `TxNotFound` path.)
- **V7-E7** (Explore) — Updated `TemplateSchema.test.ts:235` describe label to reflect the §7.1 EVSE numbering + §7.2 connector numbering split applied at V6-S1.
- **V7-E8** (Explore) — Added `part 1` qualifier to `TemplateSchema.ts:285,295` inline error messages for consistency with V6-S1 correction style.
- **V7-E9** (Explore) — Extended V6-X14 `TimeoutError` pattern uniformly: added `public override readonly name = '<ClassName>' as const` to 5 remaining error subclasses (`OCPPError`, `TemplateValidationError`, `PayloadTooLargeError`, `ConfigurationValidationError`, `AuthenticationError` — the last had a non-readonly non-`as const` override).

**Cross-agent convergence:** 8 findings flagged by 2-3 agents independently were prioritized (V7-S1+Q4, V7-S6+E4, V7-X5+Q1, V7-S2+Q7+X6). The single-agent HIGH (V7-E2) was rejected after caller-trace verification — exactly the false-positive guard added as a v7 criterion.

**Deferred (rationale documented):**
- V7-Q6 + V7-X4 — `TimeoutError` commit-body typo and "matches OCPPError pattern" inaccuracy (git history immutable; the code itself is correct).
- V7-E1 — `ConnectorStatus.transactionId: number | string` union (OCPP 1.6 numeric IDs share the field). Narrowing requires OCPP 1.6 refactor; out of scope.
- V7-E3, E6, E10, E12, E13, E14, E15, E16, E18 — verified clean by qmd or by source inspection (defensive code correct, no missing echo fields, plausible 1.6 citations not flagged as wrong).

**qmd-verified during v7** (`ocpp-specs` collection):
- §F01.FR.13 — confirmed: SHALL echo transactionId when "transaction created but not yet authorized" precondition holds; silent on status.
- TxStarted vs TxInProgress (appendix v14) — confirmed: `TxStarted` = "transaction had already started (e.g. due to cable being plugged in)" for `RequestStartTransaction` only; `TxInProgress` = generic "transaction is in progress" for `ChangeAvailability`/`Reset`/`RequestStartTransaction`.
- OCA FINAL schema for `RequestStartTransactionResponse` — confirmed: `transactionId` is an optional top-level property with no `if/then/else` constraint tied to `status`; the shape `{ status: Rejected, statusInfo: { reasonCode: TxStarted }, transactionId }` is schema-valid.
- Reference implementation comparison: `lorenzodonini/ocpp-go`, `mobilityhouse/ocpp`, `citrineos-core`, `EVerest/libocpp` — none implement F01.FR.13 (libocpp explicitly defers it with a TODO). The simulator is the only implementation found that honors the SHALL.

### V8 cross-validation review (4 agents, commit `aca724881`)

Recursive audit pass with v7 lessons elevated as criteria: regression-chain audit of v7 itself, V7-E9 blast radius verification, V7-Q1 idiom equivalence under JSON.stringify, V7-Q3 test isolation, **convergence saturation indicator** (`<3 substantive = bottom`).

**Multi-agent saturation poll:**
- Oracle1 (regression-chain): SATURATION REACHED
- Oracle2 (anti-slop): SATURATION REACHED, MERGEABILITY YES
- Librarian (external): SATURATION REACHED — 0 new commits in 4 OSS OCPP impls for F01.FR.13
- Explore (completeness): **DIMINISHING** — 3 new substantive findings found

**False positive rejected (caller-trace verification):**
- V8-E2 — Explore flagged `BaseError` as missing `name` override (would emit `name='Error'` on direct instantiation). Verified the existing constructor: `this.name = new.target.name` (line 6) already sets the correct name for any subclass AND for direct `BaseError` instantiation. False positive (same class as V6-E3, V7-E2 — runtime evidence trumps static-pattern flagging).

**3 substantive fixes applied:**
- **V8-E1 (M)** — Extended V7-E9 pattern to `ui/common` package missed entirely by v7. `ConnectionError` and `ServerFailureError` in [`ui/common/src/errors.ts`](ui/common/src/errors.ts) used mutable `this.name = '...'` constructor-assignment instead of the class-field pattern. Now all 8 error subclasses across `src/` and `ui/common` share the same `public override readonly name = '<ClassName>' as const` nominal-discriminant pattern.
- **V8-E3 (L)** — `handleRequestStartTransaction` had 1 remaining inline rejection literal at [line 2516](src/charging-station/ocpp/2.0/OCPP20IncomingRequestService.ts) (the `resolvedEvseId == null` branch with `reasonCode: NotFound`) bypassing the `buildRejectedResponse` helper used by all 11 other rejection returns in the same function. Replaced with `buildRejectedResponse(ReasonCodeEnumType.NotFound, 'No available EVSE found for remote start')` — 12/12 callsites in the start-tx handler now use the helper.
- **V8-S1 (L)** — `RequestStopTransaction` test `'should reject stop transaction for non-existent transaction ID'` used a non-UUID input string, hitting the format-check branch (`InvalidValue`) and never reaching the `TxNotFound` branch. Renamed test to `'... for invalid transaction ID format - non-UUID string'` (matches actual coverage) + added new test `'should reject stop transaction for valid UUID with no matching transaction'` using `'00000000-0000-4000-8000-000000000000'` (valid UUIDv4 format, no matching transaction). Real `TxNotFound` branch now covered.

**Verified clean during v8 (no fix needed):**
- V7-E9 blast radius: 0 sites read `error.name === '<ClassName>'`. Winston/util.inspect/JSON.stringify/Error.captureStackTrace all behave correctly (Node.js v24 empirical tests).
- V7-Q1 idiom equivalence: `&& { x }` vs `?: {}` byte-identical under `JSON.stringify` for all edge cases (`undefined`, `null`, `''`, `0`, `false`, UUID-shaped, NaN, Infinity, object).
- V7-Q3 test isolation: `beforeEach` recreates `mockStation` fresh per test; `connectorStatus.locked = true` cannot leak.
- V7-Q4 docstring trim: spec-accurate; practical anti-revert argument preserved.
- V7-Q5 spec-anchor strength: cross-reference helper + branch + test sufficient.
- Reference impl saturation: 0 new commits on F01.FR.13 across `lorenzodonini/ocpp-go`, `mobilityhouse/ocpp`, `citrineos-core`, `EVerest/libocpp`. Simulator remains the only impl that honors the SHALL.
- OCA schema unchanged 2025-09→2026-06; all v8 response shapes valid.

**Deferred (rationale documented):**
- V8-S2/Q1 — line-number pointer `~2740` in the `as UUIDv4` cast comment is fragile but the `~` already signals approximation.
- V8-S3/Q2 — §F01.FR.13 anchor at 4 sites is borderline over-anchored but each lives at a distinct cognitive level (helper API vs handler decision vs test intent); not collapsible.
- V8-E4 (echo-of-V7-S7) — `handleRequestStopTransaction` 3 inline literals return a different response type (`OCPP20RequestStopTransactionResponse`); `buildRejectedResponse` cannot be reused without a type-generic refactor.
- V8-E5 (echo) — 3 `?: {}` sites in test-helper files (not `src/`); cosmetic.

**Saturation declaration after v8:**
- 3 substantive findings applied (V8-E1, V8-E3, V8-S1). Each closes a real gap: package-coverage hole, helper-adoption hole, test-coverage hole.
- v9 sweep would target: (a) `ui/common` error subclasses (now all `readonly as const`), (b) `buildRejectedResponse` call sites (now all 12 use the helper in the start-tx handler), (c) `RequestStopTransaction` test coverage (now exercises both `InvalidValue` and `TxNotFound` branches). Expected v9 finding count: 0.

### TDZ cycle fix (commit `ac0ca2586`)

`UIServerNet.ts`, `TemplateSchema.ts` use direct leaf imports (`utils/Utils.js`) instead of the `utils/index.js` barrel, mirroring the existing pattern at `ConfigurationMigrations.ts:3`.

### Architectural rules respected

- **`src/worker/` is a standalone sub-project and must not import from elsewhere.** A `WorkerAbstract.ts` utils import added by an early audit pass was reverted in commit `2f190a4cf`. `WorkerSet.ts:191` (`JSON.stringify`) is left as-is despite being a `JSONStringify` candidate.
- The `convertToInt` site at `OCPP16IncomingRequestService.ts:785` is intentionally NOT migrated — the rejection check relies on `Number()`'s `NaN`-on-invalid contract. Inline note prevents re-introduction.

### Deferred — documented rationale

- **`secureRandom` / `sleep` / `mergeDeepRight` re-implementations in `src/worker/WorkerUtils.ts`**: worker bundle isolation rule.
- **`buildCaseInsensitiveCompositeKey` in `OCPP20VariableRegistry.ts:2517`**: NOT a duplicate of `buildConfigKey` (different argument order and key format per OCPP 2.0.1 device-model spec).
- **`handleSendMessageError` at `OCPP20RequestService.ts:131` and `OCPPResponseService.ts:89`**: base-class `sendMessage` already routes through `handleSendMessageError` at 3 sites internally; adoption would double-log.
- **`AbortSignal.timeout()` migration** (Librarian v3 finding): bigger meta-change to the `promiseWithTimeout` helper, out of scope.
- **`MapStringifyFormat` enum vs boundary-registered replacer** (Librarian v3 finding): meta-design critique of an existing helper, out of scope.
- **TS structural collapse of empty `TimeoutError` subclass** (Librarian v4 finding HC3/L1): per Q2(d) decision, location fix only; user accepted the TS narrowing risk for the single current `instanceof` site.
- **OCPP 1.6 `handleRequestGetLocalListVersion`/`handleRequestSendLocalList`** (Explore v4 finding EX1): pre-existing asymmetry with OCPP 2.0 (1.6 versions use inline catch+log+return; 2.0 versions use the helper). Out of scope for this audit PR; file a follow-up.
- **`AuthenticationError extends Error`** instead of `extends BaseError` (Explore v4 finding): pre-existing convention violation.

## Validation

- `pnpm format`
- `pnpm lint`
- `pnpm typecheck`
- `pnpm test` — **2832/2832 pass, 0 fail, 6 skipped, ~72s total**

## Diff summary

- 16 commits, 21 files, +447 / −353 net.
- Test-only changes: none.
- Behaviour-visible improvements (correctness): `getErrorMessage` emits meaningful messages for non-`Error` thrown values; `convertToBoolean` accepts more truthy representations; `JSONStringify` preserves `Map` entries previously silently dropped; `convertToInt(env.PORT)` no longer throws on unset PORT; `Number.isNaN` guard now catches non-numeric Evse keys.
- Log format change at the 10 single-catch `handleIncomingRequestError` sites in OCPP 2.0: prefix `ErrorUtils.handleIncomingRequestError: Incoming request command 'X' error:` (parity with OCPP 1.6). The 4 multi-phase nested catches in `handleRequestStartTransaction` keep the original per-phase `logger.error` for observability + use the extracted `rejectedInternalError` helper.

## Review history

- **1 review iteration with `hyperspace-insights[bot]`**: Comment 1 (real regression on `convertToInt` in OCPP16 config integer-key check) fixed by revert + empirical 4-case divergence table documented inline. Comment 2 (cosmetic loss of explicit base-10 radix in `UIServerNet.ts`) explained.
- **3 rounds of v1-style review** + **2 rounds of 4-agent cross-validation** (v3, v4): 22+ findings applied, 7+ deferred with documented rationale. All quality gates green after each round.

- `pnpm test` — **src/ 2859/2865 pass, ui/common 120/120 pass, 0 fail total, ~90s** (after v8 fixes)

## Diff summary

- 21 commits, ~28 files, +617 / −467 net (approx).  Both `src/` and `ui/common` quality gates green.
- Behaviour-visible improvements: `getErrorMessage` emits meaningful messages for non-`Error` thrown values; `convertToBoolean` accepts more truthy representations; `JSONStringify` preserves `Map` entries previously silently dropped; `convertToInt(env.PORT)` no longer throws on unset PORT; `Number.isInteger`+`< 0` guard catches non-numeric/negative Evse keys; **§F01.FR.13 spec fix: `RequestStartTransaction` rejection responses no longer fabricate `transactionId`** (11 sites in OCPP 2.0).
- Log format change at the 10 single-catch `handleIncomingRequestError` sites in OCPP 2.0: prefix `ErrorUtils.handleIncomingRequestError: Incoming request command 'X' error:` (parity with OCPP 1.6). The 4 multi-phase nested catches in `handleRequestStartTransaction` keep the original per-phase `logger.error` for observability + use the extracted `buildRejectedResponse(additionalInfo, reasonCode)` helper.

## Review history

- **1 review iteration with `hyperspace-insights[bot]`**: Comment 1 (real regression on `convertToInt` in OCPP16 config integer-key check) fixed by revert + empirical 4-case divergence table documented inline. Comment 2 (cosmetic loss of explicit base-10 radix in `UIServerNet.ts`) explained.
- **3 rounds of v1-style review** + **6 rounds of 4-agent cross-validation** (v3, v4, v5, v6, v7, v8): 47+ findings applied, 19+ deferred with documented rationale. v6 added qmd-verification as non-negotiable; v7 added false-positive caller-trace detection (V7-E2 rejected); v8 added convergence-saturation indicator + recursive fix-of-fix audit (V8-E2 rejected). Multi-round false-positive rate: 3/47 (~6%) — strict-verification criteria are paying for themselves. All quality gates green after each round.

### CODE REVIEW

File: src/charging-station/ocpp/1.6/OCPP16IncomingRequestService.ts [hunk 1]
```diff
@@ -782,7 +782,7 @@ export class OCPP16IncomingRequestService extends OCPPIncomingRequestService {
         OCPP16StandardParametersKey.WebSocketPingInterval,
       ])
       if (integerKeys.has(keyToChange.key as OCPP16StandardParametersKey)) {
-        const numValue = Number(value)
+        const numValue = convertToInt(value)
```
[the remark this answers is not in the document]
Comment (jerome-benoit) [reply src/charging-station/ocpp/1.6/OCPP16IncomingRequestService.ts]: Valid catch — and actually broader than just `null`/`undefined`. Empirically verified divergence:

| value | `Number(value)` | `convertToInt(value)` |
|---|---|---|
| `undefined` | `NaN` | `0` (silently accepted) |
| `''` | `0` | **throws uncaught** |
| `'1.5'` | `1.5` (rejected) | `1` (silently accepted — truncation) |
| `'abc'` | `NaN` | **throws uncaught** |

Two cases produce uncaught exceptions that propagate out of `handleRequestChangeConfiguration` instead of returning Rejected; the other two silently accept invalid values. The rejection check `!Number.isInteger(n) || n < 0` relies on `Number()`'s NaN-on-invalid contract.

Fixed in `893968ad1` — reverted this site to `Number(value)` with an inline note to prevent re-introduction by future cleanups. The suggested `convertToInt(value ?? '')` was not used because `convertToInt('')` throws (would make it worse).

File: src/charging-station/ui-server/UIServerNet.ts [hunk 2]
```diff
@@ -94,7 +99,7 @@ const isValidPort = (port: string | undefined): boolean => {
   if (!/^\d+$/.test(port)) {
     return false
   }
-  const parsedPort = Number.parseInt(port, 10)
+  const parsedPort = convertToInt(port)
```
[the remark this answers is not in the document]
Comment (jerome-benoit) [reply src/charging-station/ui-server/UIServerNet.ts:100]: Technically correct — `convertToInt` internally calls `Number.parseInt(value)` without an explicit radix, which would interpret `'0xff'` as hex (`255`).

However at this specific call site the preceding regex `/^\d+$/` on line 95 rejects any input containing a non-decimal character before `convertToInt` is ever invoked. The radix loss is unobservable here:

```typescript
if (!/^\d+$/.test(port)) {     // ← '0xff', '0o17', '1e5' all rejected here
  return false
}
const parsedPort = convertToInt(port)
```

Keeping the `convertToInt` adoption for consistency with the audit. If the broader concern is `convertToInt` itself silently accepting hex-prefixed strings, that would be a fix to `Utils.ts:266` (`Number.parseInt(value)` → `Number.parseInt(value, 10)`) — outside this PR's scope.
